File: jvm/ship-conquest/src/main/kotlin/com/example/shipconquest/domain/ship_navigation/ShipPath.kt
package com.example.shipconquest.domain.ship_navigation

import com.example.shipconquest.domain.Position
import com.example.shipconquest.domain.toPosition
import java.time.Duration
import java.time.Instant

class ShipPath(
    val landmarks: List<CubicBezier>,
    val startTime: Instant,
    val duration: Duration
) {
    val endTime: Instant = startTime + duration

    fun getPosition(u: Double): Position {
        val index = u.toInt()
        if (index >= landmarks.size) return landmarks.last().p3.toPosition()

        val n = u % 1
        return landmarks[index].get(n)
    }

    fun getPositionFromTime(): Position {
        val percentage = getTimePercentage(startTime, endTime)
        return getPosition(percentage * landmarks.size)
    }

    fun getCurrentDuration(): Duration = Duration.between(endTime, Instant.now())

    private fun getTimePercentage(startTime: Instant, endTime: Instant): Double {
        val now = Instant.now()
        val totalDuration = Duration.between(startTime, endTime).toMillis()
        val elapsedDuration = Duration.between(startTime, now).toMillis()
        return elapsedDuration.toDouble() / totalDuration.toDouble()
    }
}

File: flutter/lib/domain/ship/utils/classes/ship_path.dart
import 'dart:math';

import 'package:ship_conquest/domain/space/cubic_bezier.dart';
import 'package:ship_conquest/domain/space/position.dart';
import 'package:ship_conquest/domain/utils/get_time_percentage.dart';

/// Represents a ship Path
/// this data is global and reliable throughout the whole path.
class ShipPath {
  // path landmarks/points
  final List<CubicBezier> landmarks;
  // path start time
  final DateTime startTime;
  // path travelling duration
  final Duration duration;
  // constructor
  ShipPath({required this.landmarks, required this.startTime, required this.duration});
  // optimizations
  late final endTime = startTime.add(duration);

  /// Retrieves a [Position] according to [u]
  Position getPosition(double u) {
    final index = u.floor();
    if (index == landmarks.length) return landmarks.last.p3 * 1.0;

    final n = u % 1;
    return landmarks[index].get(n);
  }

  /// Retrieves a the current ship [Position].
  Position getPositionFromTime() {
    final percentage = getTimePercentage(startTime, endTime);
    return getPosition(percentage * landmarks.length);
  }

  /// Retrieves the ship's angle according to its interpolation value [u].
  double getAngle(double u) {
    final index = u.floor();
    if (index == landmarks.length) return landmarks.last.getAngleAtPoint(1.0);

    final n = u % 1;
    return landmarks[index].getAngleAtPoint(n);
  }

  /// Retrieves the ship's angle according to the time.
  double getAngleFromTime() {
    final percentage = getTimePercentage(startTime, endTime);
    return getAngle(percentage * landmarks.length);
  }

  /// Gets the value from the boat's path position, or the landmarks length value.
  double getStartFromTime() {
    final timeElapsed = DateTime.now().difference(startTime);
    final percentage = timeElapsed.inMilliseconds / duration.inMilliseconds;
    return min(landmarks.length.toDouble(), percentage * landmarks.length);
  }

  /// Retrieves the duration from the current time to the end of the path.
  Duration getCurrentDuration() => endTime.difference(DateTime.now());

  /// Checks if the ship has reached its destination by comparing the
  /// path's end time to the current instant.
  bool hasReachedDestiny() => endTime.isBefore(DateTime.now());
}